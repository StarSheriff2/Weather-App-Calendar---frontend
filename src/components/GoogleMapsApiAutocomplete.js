/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable react/destructuring-assignment */
/* eslint-disable no-console */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng,
} from 'react-places-autocomplete';
import { v4 } from 'uuid';

export default class GoogleMapsApiAutocomplete extends Component {
  constructor(props) {
    super(props);

    this.handleChange = this.handleChange.bind(this);
    this.handleSelect = this.handleSelect.bind(this);
  }

  handleChange(address) {
    this.props.form.setValues({ ...this.props.form.values, city: address });
  }

  handleSelect(address) {
    geocodeByAddress(address)
      .then((results) => getLatLng(results[0]))
      .then((latLng) => {
        this.props.form.setValues({ ...this.props.form.values, city: address, locationCoordinates: `${latLng.lat.toString()}, ${latLng.lng.toString()}` });
      })
      .catch((error) => console.error(error));
  }

  render() {
    console.log('form prop: ', this.props.form);
    const { name, value } = this.props.field;
    const { placeholder } = this.props;

    return (
      <PlacesAutocomplete
        value={value}
        onChange={this.handleChange}
        onSelect={this.handleSelect}
      >
        {({
          getInputProps, suggestions, getSuggestionItemProps, loading,
        }) => (
          <div>
            <input
              {...getInputProps({
                placeholder,
                className: 'location-search-input form-control',
                name,
              })}
            />
            <div className="autocomplete-dropdown-container">
              {loading && <div>Loading...</div>}
              {suggestions.map((suggestion) => {
                const className = suggestion.active
                  ? 'suggestion-item--active'
                  : 'suggestion-item';
                // inline style for demonstration purpose
                const style = suggestion.active
                  ? { backgroundColor: '#fafafa', cursor: 'pointer' }
                  : { backgroundColor: '#ffffff', cursor: 'pointer' };
                return (
                  <div
                    key={v4()}
                    {...getSuggestionItemProps(suggestion, {
                      className,
                      style,
                    })}
                  >
                    <span>{suggestion.description}</span>
                  </div>
                );
              })}
            </div>
          </div>
        )}
      </PlacesAutocomplete>
    );
  }
}

GoogleMapsApiAutocomplete.propTypes = {
  field: PropTypes.shape({
    name: PropTypes.string.isRequired,
    value: PropTypes.string,
  }).isRequired,
  form: PropTypes.shape(
    null,
  ).isRequired,

  placeholder: PropTypes.string.isRequired,
};
