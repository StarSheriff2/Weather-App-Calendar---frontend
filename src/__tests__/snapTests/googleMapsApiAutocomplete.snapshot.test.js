/* eslint-disable max-classes-per-file */
/* eslint-disable class-methods-use-this */
import React from 'react';
import { cleanup } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import renderWithRedux from '../../tests/test-utils';
import GoogleMapsApiAutocomplete from '../../components/GoogleMapsApiAutocomplete';

afterEach(cleanup);

describe('GoogleMapsApiAutocomplete', () => {
  // ⬇️ window.google object code snippet copied from this thread:
  // https://github.com/hibiken/react-places-autocomplete/issues/189#issuecomment-418555543

  window.google = {
    maps: {
      Marker: class {},
      Map: class {
        setTilt() {}

        fitBounds() {}
      },
      LatLngBounds: class {},
      places: {
        Autocomplete: class {},
        AutocompleteService: class {},
        PlacesServiceStatus: {
          INVALID_REQUEST: 'INVALID_REQUEST',
          NOT_FOUND: 'NOT_FOUND',
          OK: 'OK',
          OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
          REQUEST_DENIED: 'REQUEST_DENIED',
          UNKNOWN_ERROR: 'UNKNOWN_ERROR',
          ZERO_RESULTS: 'ZERO_RESULTS',
        },
        PlacesAutocomplete: {
          INVALID_REQUEST: 'INVALID_REQUEST',
          NOT_FOUND: 'NOT_FOUND',
          OK: 'OK',
          OVER_QUERY_LIMIT: 'OVER_QUERY_LIMIT',
          REQUEST_DENIED: 'REQUEST_DENIED',
          UNKNOWN_ERROR: 'UNKNOWN_ERROR',
          ZERO_RESULTS: 'ZERO_RESULTS',
        },
      },

      MarkerClusterer: class {},
      Geocoder: class {},
    },
  };

  test('Renders the Component', () => {
    const field = {
      name: 'city',
      value: '',
    };

    const placeholder = 'Search location';

    const form = {
      values: {
        description: '',
        date: '',
        time: `${new Date().toString().slice(16, 21)}`,
        city: '',
        locationCoordinates: '',
      },
      setValues: () => true,
    };

    const googleMapsApiAutocomplete = renderWithRedux(
      <Router>
        <GoogleMapsApiAutocomplete
          field={field}
          placeholder={placeholder}
          form={form}
        />
      </Router>,
    ).toJSON();

    expect(googleMapsApiAutocomplete).toMatchSnapshot();
  });
});
