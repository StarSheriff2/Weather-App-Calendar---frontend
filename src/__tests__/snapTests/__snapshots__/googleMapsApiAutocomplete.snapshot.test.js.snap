// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoogleMapsApiAutocomplete Renders the Component 1`] = `
<div>
  <input
    aria-autocomplete="list"
    aria-expanded={false}
    autoComplete="off"
    className="location-search-input form-control"
    disabled={false}
    name="city"
    onBlur={[Function]}
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="Search location"
    role="combobox"
    type="text"
    value=""
  />
  <div
    className="autocomplete-dropdown-container"
  />
</div>
`;
